### PR TITLE
Added test, pytest configuration, removed unnecessary import

### LINKED ISSUE #34: Adding unittests
There are not really tests in this repository which makes it a bit hard to ensure there are no bugs and/or unintentional behavior.

To fix this, unittests should be written.
When this is done, integration tests can be done too.

~~Right now there is only one (unnecessary) test.~~
All new tests should be created in a new directory called `tests/` and because it is already in use, `pytest` should be used as test-framework.

### PR DESCRIPTION
**Pull request by GrbavaCigla**
Resolves #34

### CODE REVIEW

**Comment by GrbavaCigla** [comment]
Please let me know if you have an idea for more tests.

File: src/tests/__init__.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+from src.Board import Board
+from src.InputParser import InputParser
+
+def makeBoardMoves(board: Board, moves: List[str]):
+    parser = InputParser(board, True)
+    
+    for moveStr in moves:
+        move = parser.parse(moveStr)
+        parser.side = not parser.side
+        board.makeMove(move)
```
Comment (ClasherKasten) [inline src/tests/__init__.py:12]: `__init__.py` files shouldn't contain any code. Please move this to `conftest.py` and make it a fixture.

File: src/tests/__init__.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+from typing import List
```
Comment (ClasherKasten) [inline src/tests/__init__.py:1]: Don't type annotate tests. It is to much work. if mypy is not happy with that we will tell mypy to not check the tests.

File: pytest.ini [hunk 3]
```diff
@@ -0,0 +1,4 @@
+[pytest]
+addopts = -rP
+testpaths = src/tests
```
Comment (ClasherKasten) [inline pytest.ini:3]: Never create tests inside the `src` directory.

**Review by ClasherKasten — CHANGES_REQUESTED** [review]
Round 1:
- Most important: move the tests into a separate folder named `tests/`
- For the rest take a look at the comments I made.

If you have any questions, just ask.

**Comment by ClasherKasten** [comment]
> Please let me know if you have an idea for more tests.

Ideally I want to have a test coverage of 100%.
So simply the idea would be to write tests for every part of the code that isn't covered by the tests. This doesn't have to be achieved with one PR, but this is the goal.
Please tell me if you want to push further commits that add new tests to this PR or if I should merge it after the changes and you open more PRs for new tests (but please don't open a new PR for every single test).

**Comment by GrbavaCigla** [comment]
Thanks for the review.

Also why `main.py` instead of `__main__.py`?

**Comment by ClasherKasten** [comment]
> Also why `main.py` instead of `__main__.py`?

`main.py` is just a normal python module and can be used like one. `__main__.py` has a special meaning for python. I think simplified `__main__.py` turns a directory into an executable python module.
As an example suppose you have a directory called `some_dir` with an `__main__.py` file which contains all the startup code for the application. You could then use `python -m some_dir` to run you application.

File: tests/conftest.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+from src.Board import Board
+from src.InputParser import InputParser
+
+def makeBoardMoves(board: Board, moves):
+    parser = InputParser(board, True)
+    
+    for moveStr in moves:
+        move = parser.parse(moveStr)
+        parser.side = not parser.side
+        board.makeMove(move)
```
Comment (ClasherKasten) [inline tests/conftest.py:10]: Eventually i should've marked it a bit clearer. This should be converted to a fixture. `conftest.py` has a special meaning to `pytest` (it is kind of a setup file), it will be automatically **"run"** before the actual tests run.

**Review by ClasherKasten — CHANGES_REQUESTED** [review]
Round 2: I already told you that in the last Round.

File: tests/conftest.py (hunk 4 above)
Comment (GrbavaCigla) [reply tests/conftest.py:10]: Ah, ok, sorry.

**Comment by GrbavaCigla** [comment]
> > Also why `main.py` instead of `__main__.py`?
> 
> `main.py` is just a normal python module and can be used like one. `__main__.py` has a special meaning for python. I think simplified `__main__.py` turns a directory into an executable python module. As an example suppose you have a directory called `some_dir` with an `__main__.py` file which contains all the startup code for the application. You could then use `python -m some_dir` to run you application.

Exactly, shouldn't this module be executable since project is cli chess?

Anyways, I will fix this tomorrow since it is getting late.

**Comment by ClasherKasten** [comment]
> Exactly, shouldn't this module be executable since project is cli

If you would've taken a look into `README.md` in the sections **Installation** and **Usage** it is written down how to properly execute it. But yeah, a `__main__.py` is an option to. So if you want to add, feel free to do so.

**Comment by GrbavaCigla** [comment]
Can you tell me what functions should I unittest because in `testMate` I am testing `isCheckmate`?

**Comment by ClasherKasten** [comment]
> Can you tell me what functions should I unittest because in `testMate` I am testing `isCheckmate`?

Here is a little definition:
> A unit test should have no dependencies on code outside the unit tested. You decide what the unit is by looking for the smallest testable part. Where there are dependencies they should be replaced by false objects.
[StackOverflow](https://stackoverflow.com/a/14186648/13891412)

I think there is a way to isolate the checkmate behavior much better from the rest of the program. And if that's not something you want to do right now, there are enough other parts of the code that need to be tested as well. As an example, `Coordinate` is completely isolated. Most of the types of pieces can be tested when the board is mocked (or the test for the board is also pushed, or if you have another solution just propose it).

What I'd like to say in conclusion is that it's fine if you want to verify that it actually analyzes, moves and recognizes everything (the whole process and how each small part is put together). But then you shouldn't do it just for checkmate. Just checking that it validates a specific example is not the idea of tests. Tests should cover as many possibilities as possible (which includes edge cases). And currently I don't really see this code covering anything (especially not edge cases).

**Comment by GrbavaCigla** [comment]
Yeah, added more tests, but it isn't easy writing 'isolated' tests while everything is entangled in code.

**Comment by ClasherKasten** [comment]
> Yeah, added more tests, but it isn't easy writing 'isolated' tests while everything is entangled in code.

That's a valid point. So you identified that the problem is in the code. A solution to this could also be to discuss how the code could be changed to be more isolated.

But I don't wanna be so negative. Instead: the new tests look very good and seem to improve coverage quite a bit. 👍
Still I would like an answer if you wanna push more or if I can merge this PR and new ones can be opened for more tests?

**Comment by GrbavaCigla** [comment]
You can merge it :D
